KAFKA-3361: Initial protocol documentation page and generation

granthenke opened the pull request:
- Moves all generated docs under /docs/generated
- Generates docs for Protocol, Errors, and ApiKeys
- Adds new protocol.html page

jkreps commented:
This is a super cool idea.

christian-posta commented:
This is great. Where's the tool you used to generate this? Can it do markdown?

Would love to work on converting the Kafka main-page docs to gitbook.com format https://www.gitbook.com

granthenke commented:
@christian-posta No tool was used. All the code used to generate this is in this pull request. Just some hacked together code to build the output.

granthenke committed:
KAFKA-3361: Initial protocol documentation page and generation

- Moves all generated docs under /docs/generated
- Generates docs for Protocol, Errors, and ApiKeys
- Adds new protocol.html page

granthenke commented:
Below was my original WIP message:

> Please see the email with the subject "Protocol Documentation WIP" in the Kafka dev mailing list for context on this PR. Do not review this for merge in its current state. 
> 
> Here are links for the generated output:
> - Gist of output: https://gist.github.com/granthenke/dae9e60b5e548acead58#file-kafka_protocol-html
> - In browser preview of Gist: https://rawgit.com/granthenke/dae9e60b5e548acead58/raw/02e93fae37381fbbeaf83d55e4b4c4083f87319b/kafka_protocol.html

I have updated the patch to be a "real" patch. The protocol docs still need some work. But that work is tracked by the umbrella jira here: [KAFKA-3360](https://issues.apache.org/jira/browse/KAFKA-3360).

granthenke commented:
@ewencp Ping for review. Thoughts on how this should link into the main docs?

gwenshap commented:
I agree that this should be a separate page (until we break down our documentation).

How about: Add a link to the side-bar (top.html) that will point to this page directly?

gwenshap commented on docs/protocol.html:
```diff
@@ -0,0 +1,163 @@
+<h3><a id="protocol" href="#protocol">Kafka Wire Protocol</a></h3>
+
+<p>This document covers the protocol implemented in Kafka 0.8 and beyond. It is meant to give a readable guide to the protocol that covers the available requests, their binary format, and the proper way to make use of them to implement a client. This document assumes you understand the basic design and terminology described <a href="https://kafka.apache.org/documentation.html#design">here</a></p>
```
Since this page is now in the version control and generated, I'm not sure this makes sense... we'll have a separate generated doc with each release, documenting the protocol for current version.

granthenke commented:
@gwenshap Good idea. The side-bar is probably the best place. I can send a PR to the site repo after this gets in. 

I updated the PR to remove the version.

granthenke committed:
Remove version

gwenshap commented:
LGTM (both the code and the plan).

Also, we want to publish the protocol for 0.9 too, right? So we'll want to cherrypick this to the 0.9 branch.

gwenshap commented:
Sorry, this doesn't cherrypick cleanly. 
So, either we do this for 0.10.0 and up (in which case your site-repo PR will have the honor of creating the 0.10 docs directory), or you can do a separate PR for 0.9.0. Either way is fine.

granthenke commented:
@gwenshap I noticed a copy/paste bug while backporting. That fix for trunk is PR #1037. The backport PR for the 0.9 branch contains the fix already. Its PR #1038.
